File: samples/C/diffusion/diffusion.c


#include <math.h>
#include <stdio.h>

#include <solomon.hpp>

#ifndef M_PI
#define M_PI 3.14159265358979323846264338328  // pi
#endif                                        // M_PI

#define INDEX(nx, ny, nz, ii, jj, kk) ((kk) + (nz) * ((jj) + (ny) * (ii)))

#define IMIN(a, b) (((a) < (b)) ? (a) : (b))
#define IMAX(a, b) (((a) > (b)) ? (a) : (b))

double diffusion3d(int nx, int ny, int nz, float dx, float dy, float dz, float dt, float kappa,
                   const float *restrict f, float *restrict fn) {
  const float ce = kappa * dt / (dx * dx);
  const float cw = ce;
  const float cn = kappa * dt / (dy * dy);
  const float cs = cn;
  const float ct = kappa * dt / (dz * dz);
  const float cb = ct;

  const float cc = 1.0 - (ce + cw + cn + cs + ct + cb);

  OFFLOAD(AS_INDEPENDENT, COLLAPSE(3), ACC_CLAUSE_PRESENT(f, fn))
  for (int i = 0; i < nx; i++) {
    for (int j = 0; j < ny; j++) {
      for (int k = 0; k < nz; k++) {
        const int ix = INDEX(nx, ny, nz, i, j, k);
        const int ip = INDEX(nx, ny, nz, IMIN(i + 1, nx - 1), j, k);
        const int im = INDEX(nx, ny, nz, IMAX(i - 1, 0), j, k);
        const int jp = INDEX(nx, ny, nz, i, IMIN(j + 1, ny - 1), k);
        const int jm = INDEX(nx, ny, nz, i, IMAX(j - 1, 0), k);
        const int kp = INDEX(nx, ny, nz, i, j, IMIN(k + 1, nz - 1));
        const int km = INDEX(nx, ny, nz, i, j, IMAX(k - 1, 0));

        fn[ix] = cc * f[ix] + ce * f[ip] + cw * f[im] + cn * f[jp] + cs * f[jm] + ct * f[kp] + cb * f[km];
      }
    }
  }

  return (double)(nx * ny * nz) * 13.0;
}

void init(int nx, int ny, int nz, float dx, float dy, float dz, float *f) {
  const float kx = 2.0F * (float)M_PI;
  const float ky = kx;
  const float kz = kx;

  for (int i = 0; i < nx; i++) {
    for (int j = 0; j < ny; j++) {
      for (int k = 0; k < nz; k++) {
        const int ix = INDEX(nx, ny, nz, i, j, k);
        const float x = dx * ((float)i + 0.5F);
        const float y = dy * ((float)j + 0.5F);
        const float z = dz * ((float)k + 0.5F);

        f[ix] = 0.125F * (1.0F - cosf(kx * x)) * (1.0F - cosf(ky * y)) * (1.0F - cosf(kz * z));
      }
    }
  }
}

double accuracy(double time, int nx, int ny, int nz, float dx, float dy, float dz, float kappa, const float *f) {
  const float kx = 2.0F * (float)M_PI;
  const float ky = kx;
  const float kz = kx;

  const float ax = expf(-kappa * (float)time * (kx * kx));
  const float ay = expf(-kappa * (float)time * (ky * ky));
  const float az = expf(-kappa * (float)time * (kz * kz));

  double ferr = 0.0;

  for (int i = 0; i < nx; i++) {
    for (int j = 0; j < ny; j++) {
      for (int k = 0; k < nz; k++) {
        const int ix = INDEX(nx, ny, nz, i, j, k);
        const float x = dx * ((float)i + 0.5F);
        const float y = dy * ((float)j + 0.5F);
        const float z = dz * ((float)k + 0.5F);

        const float f0 = 0.125F * (1.0F - ax * cosf(kx * x)) * (1.0F - ay * cosf(ky * y)) * (1.0F - az * cosf(kz * z));

        ferr += (f[ix] - f0) * (f[ix] - f0);
      }
    }
  }

  return sqrt(ferr / (double)(nx * ny * nz));
}

File: samples/F/diffusion/diffusion.f90
#include <solomon.hpp>

#ifndef M_PI
#define M_PI 3.14159265358979323846264338328_dp
#endif

module diffusion
  USE_SOLOMON_RUNTIME
  use misc
  implicit none

contains

  function diffusion3d(nx, ny, nz, dx, dy, dz, dt, kappa, f, fn)
    integer,  intent(in)  :: nx, ny, nz
    real(sp), intent(in)  :: dx, dy, dz, dt, kappa
    real(sp), intent(in)  :: f(nx,ny,nz)
    real(sp), intent(out) :: fn(nx,ny,nz)
    real(sp) :: ce,cw,cn,cs,ct,cb,cc
    integer :: i,j,k,ip,im,jp,jm,kp,km
    integer :: w,e,n,s,b,t
    real(dp) :: diffusion3d

    ce = kappa*dt/(dx*dx)
    cw = ce
    cn = kappa*dt/(dy*dy)
    cs = cn
    ct = kappa*dt/(dz*dz)
    cb = ct

    cc = 1.0 - (ce + cw + cn + cs + ct + cb)

    SOLOMON_OFFLOAD(SOLOMON_CLAUSE_INDEPENDENT, SOLOMON_CLAUSE_COLLAPSE(3), ACC_CLAUSE_PRESENT(f, fn), \
            SOLOMON_CLAUSE_PRIVATE(i,j,k,ip,im,jp,jm,kp,km,w,e,n,s,b,t))
    do k=1, nz
       do j=1, ny
          do i=1, nx
             ip = MERGE(i+1,i,i+1<=nx)
             im = MERGE(i-1,i,i-1>= 1)
             jp = MERGE(j+1,j,j+1<=ny)
             jm = MERGE(j-1,j,j-1>= 1)
             kp = MERGE(k+1,k,k+1<=nz)
             km = MERGE(k-1,k,k-1>= 1)

             fn(i,j,k) = cc * f(i,j,k) &
                  + ce * f(ip,j,k) + cw * f(im,j,k) &
                  + cn * f(i,jp,k) + cs * f(i,jm,k) &
                  + ct * f(i,j,kp) + cb * f(i,j,km)
          end do ! i
       end do ! j
    end do ! k
    SOLOMON_END_OFFLOAD

    diffusion3d = real(nx*ny*nz,dp)*13.0_dp
  end function diffusion3d


  subroutine init(nx, ny, nz, dx, dy, dz, f)
    integer,  intent(in)  :: nx, ny, nz
    real(sp), intent(in)  :: dx, dy, dz
    real(sp), intent(out) :: f(nx,ny,nz)
    real(sp) :: kx,ky,kz,x,y,z
    integer  :: i,j,k

    kx = 2.0_sp * real(M_PI,sp)
    ky = kx
    kz = kx

    do k=1, nz
       do j = 1, ny
          do i = 1, nx
             x = dx*(real(i-1,sp) + 0.5_sp)
             y = dy*(real(j-1,sp) + 0.5_sp)
             z = dz*(real(k-1,sp) + 0.5_sp)

             f(i,j,k) = 0.125_sp &
                  * (1.0_sp - cos(kx*x)) &
                  * (1.0_sp - cos(ky*y)) &
                  * (1.0_sp - cos(kz*z))
          end do ! i
       end do ! j
    end do ! k
  end subroutine init


  function accuracy(time, nx, ny, nz, dx, dy, dz, kappa, f)
    real(dp), intent(in)  :: time
    integer,  intent(in)  :: nx, ny, nz
    real(sp), intent(in)  :: dx, dy, dz, kappa
    real(sp), intent(in)  :: f(nx,ny,nz)
    real(sp) :: kx,ky,kz,ax,ay,az,x,y,z,f0
    real(dp) :: ferr
    integer  :: i,j,k
    real(dp) :: accuracy

    kx = 2.0_sp*real(M_PI,sp)
    ky = kx
    kz = kx

    ax = exp(-kappa*real(time,sp)*(kx*kx))
    ay = exp(-kappa*real(time,sp)*(ky*ky))
    az = exp(-kappa*real(time,sp)*(kz*kz))

    ferr = 0.0_dp

    do k=1, nz
       do j=1, ny
          do i=1, nx
             x = dx*(real(i-1,sp) + 0.5_sp)
             y = dy*(real(j-1,sp) + 0.5_sp)
             z = dz*(real(k-1,sp) + 0.5_sp)

             f0 = 0.125_sp &
                  * (1.0_sp - ax*cos(kx*x)) &
                  * (1.0_sp - ay*cos(ky*y)) &
                  * (1.0_sp - az*cos(kz*z))

             ferr = ferr + real(f(i,j,k) - f0,dp)**2
          end do ! i
       end do ! j
    end do ! k

    accuracy = sqrt(ferr/real(nx*ny*nz,dp))
  end function accuracy

end module diffusion
